lib/tasks/emails.rake

```diff
@@ -0,0 +1,94 @@
+namespace :emails do
+  desc "Count the number of users who have received a survey"
+  task :count_2021_03_survey_recipients, %i[] => [:environment] do |_, _args|
+    survery_count = User.where(has_received_2021_03_survey: true).count
+    puts "Number of 2021_03_survey already sent: #{survery_count}"
+  end
+
+  desc "send test email"
+  task :test_email, %i[email] => [:environment] do |_, _args|
+    abort("Please provide an email") if args.email.nil?
+
+    UserMailer.with(
+      email: args.email,
+      subject: default_email_subject_line,
+      body: default_email_body,
+    ).adhoc_email.deliver_later
+  end
+
+  desc "Send a survey to user groups in cohorts of most recently logged in"
+  task :send_2021_03_survey_by_login_cohorts,
+       %i[ten_minute_group_limit one_minute_group_limit remaining_group_limit] => [:environment] do |_, args|
+    args.with_defaults(
+      ten_minute_group_limit: 50,
+      one_minute_group_limit: 100,
+      remaining_group_limit: 350,
+    )
+
+    ten_minute_group = User.where(feedback_consent: true, has_received_2021_03_survey: false)
+                           .where("last_sign_in_at > created_at + interval '10 minute'")
+                           .order(Arel.sql("RANDOM()"))
+                           .limit(args.ten_minute_group_limit)
+
+    puts "10 minute group: #{ten_minute_group.count}"
+
+    one_minute_group = User.where(feedback_consent: true, has_received_2021_03_survey: false)
+                           .where("last_sign_in_at > created_at + interval '1 minute'")
+                           .where.not(id: ten_minute_group.pluck(:id))
+                           .order(Arel.sql("RANDOM()"))
+                           .limit(args.one_minute_group_limit)
+
+    puts "1 minute group: #{one_minute_group.count}"
+
+    remaining_group = User.where(feedback_consent: true, has_received_2021_03_survey: false)
+                          .where.not(id: ten_minute_group.pluck(:id))
+                          .where.not(id: one_minute_group.pluck(:id))
+                          .order(Arel.sql("RANDOM()"))
+                          .limit(args.remaining_group_limit)
+
+    puts "Remaining group: #{remaining_group.count}"
```

Comment:
Non-blocking suggestion: I don't know if there's a limit on `WHERE NOT IN` clauses, but if we're going to send this batch to thousands, maybe this would be better:

```ruby
    ten_minute_group = User.where(feedback_consent: true, has_received_2021_03_survey: false)
                           .where("last_sign_in_at > created_at + interval '10 minute'")
                           .order(Arel.sql("RANDOM()"))
                           .limit(args.ten_minute_group_limit)
    ten_minute_group.update_all(has_received_2021_03_survey: true)

    puts "10 minute group: #{ten_minute_group.count}"

    one_minute_group = User.where(feedback_consent: true, has_received_2021_03_survey: false)
                           .where("last_sign_in_at > created_at + interval '1 minute'")
                           .order(Arel.sql("RANDOM()"))
                           .limit(args.one_minute_group_limit)
    one_minute_group.update_all(has_received_2021_03_survey: true)

    puts "1 minute group: #{one_minute_group.count}"

    remaining_group = User.where(feedback_consent: true, has_received_2021_03_survey: false)
                          .order(Arel.sql("RANDOM()"))
                          .limit(args.remaining_group_limit)
    remaining_group.update_all(has_received_2021_03_survey: true)

    puts "Remaining group: #{remaining_group.count}"
```

Reply:
Ah i see!
Yeah interesting 🤔 I'm cautious about changing the DB state before we've tried to end the email.
Not sure there's much that could go wrong there... but I wouldn't want to loose that state if UserMailer packed up before we sent stuff.

I'll see how this batch goes?